src/vs/workbench/browser/parts/editor/editorParts.ts

```diff
@@ -679,27 +708,72 @@ export class EditorParts extends MultiWindowParts<EditorPart> implements IEditor
 		};
 	}
 
-	private updateGlobalContextKeys(): void {
-		const activeGroupScopedContextKeys = this.scopedContextKeys.get(this.activeGroup.id);
-		if (!activeGroupScopedContextKeys) {
-			return;
+	private readonly contextKeyProviders = new Map<string, IEditorGroupContextKeyProvider<ContextKeyValue>>();
+	private readonly registeredContextKeys = new Map<GroupIdentifier, Map<string, IContextKey>>();
+
+	registerContextKeyProvider<T extends ContextKeyValue>(provider: IEditorGroupContextKeyProvider<T>): IDisposable {
+		if (this.contextKeyProviders.has(provider.contextKey.key) || this.globalContextKeys.has(provider.contextKey.key)) {
+			throw new Error(`A context key provider for key ${provider.contextKey.key} already exists.`);
 		}
 
-		for (const [key, globalContextKey] of this.globalContextKeys) {
-			const scopedContextKey = activeGroupScopedContextKeys.get(key);
-			if (scopedContextKey) {
-				globalContextKey.set(scopedContextKey.get());
-			} else {
-				globalContextKey.reset();
+		this.contextKeyProviders.set(provider.contextKey.key, provider);
+
+		const setContextKeyForGroups = () => {
+			for (const group of this.groups) {
+				this.runRegisteredContextKeyProvider(group, provider);
 			}
-		}
+		};
+
+		// Run initially and on change
+		setContextKeyForGroups();
+		const onDidChange = provider.onDidChange?.(() => setContextKeyForGroups());
+
+		return toDisposable(() => {
+			onDidChange?.dispose();
+			this.contextKeyProviders.delete(provider.contextKey.key);
+			this.globalContextKeys.delete(provider.contextKey.key);
+			this.scopedContextKeys.forEach(scopedContextKeys => scopedContextKeys.delete(provider.contextKey.key));
+			this.registeredContextKeys.forEach(registeredContextKeys => registeredContextKeys.delete(provider.contextKey.key));
+		});
 	}
 
-	private removeGroupScopedContextKeys(group: IEditorGroupView): void {
-		const groupScopedContextKeys = this.scopedContextKeys.get(group.id);
-		if (groupScopedContextKeys) {
-			this.scopedContextKeys.delete(group.id);
+	private readonly groupContextKeyProvidersRunners = this._register(new DisposableMap<GroupIdentifier, IDisposable>());
+	private registerRunGroupContextKeyProviders(group: IEditorGroupView): void {
```

Comment:
This method is only called when a group is added, but aren't we missing to run it initially for when a context key provider is registered on the groups?

Reply:
No that is already handled when registering it. When a provider is registered, it runs the provider for each group. 
`registerRunGroupContextKeyProviders` however makes sure all providers of one group are run for which the active editor has changed